File: src/application/TaskService.ts
/**
 * Casos de uso RF01–RF08 — camada de aplicação.
 */

import { TarefaNaoEncontradaError } from "../domain/errors.js";
import {
  Task,
  type PrioridadeTarefa,
  type StatusTarefa,
} from "../domain/Task.js";
import type { TaskRepository } from "./TaskRepository.js";

export type FiltroTarefa = "todas" | "pendentes" | "concluidas";

export class TaskService {
  constructor(private readonly repository: TaskRepository) {}

  async inicializar(): Promise<void> {
    await this.repository.carregar();
  }

  /** RF01 */
  async criarTarefa(
    titulo: string,
    prioridade?: PrioridadeTarefa
  ): Promise<Task> {
    const tarefa = Task.criar(titulo, prioridade);
    this.repository.salvar(tarefa);
    await this.repository.persistir();
    return tarefa;
  }

  /** RF02 + RF07 + RN08 */
  listarTarefas(filtro: FiltroTarefa = "todas"): Task[] {
    const todas = this.repository.listarTodas();
    if (filtro === "pendentes") {
      return todas.filter((t) => t.status === "Pendente");
    }
    if (filtro === "concluidas") {
      return todas.filter((t) => t.status === "Concluída");
    }
    return todas;
  }

  /** RF03 */
  async concluirTarefa(id: string): Promise<Task> {
    const tarefa = this.obterOuFalhar(id);
    if (tarefa.status !== "Concluída") {
      tarefa.concluir();
      this.repository.atualizar(tarefa);
      await this.repository.persistir();
    }
    return tarefa;
  }

  /** RF04 + RF06 */
  async editarTarefa(
    id: string,
    dados: {
      titulo?: string;
      prioridade?: PrioridadeTarefa;
      status?: StatusTarefa;
    }
  ): Promise<Task> {
    const tarefa = this.obterOuFalhar(id);
    tarefa.atualizar(dados);
    this.repository.atualizar(tarefa);
    await this.repository.persistir();
    return tarefa;
  }

  /** RF05 */
  async excluirTarefa(id: string): Promise<void> {
    this.obterOuFalhar(id);
    this.repository.excluir(id);
    await this.repository.persistir();
  }

  /** Utilitário para testes — limpa persistência. */
  async resetarTodas(): Promise<void> {
    const repo = this.repository as TaskRepository & { limpar?: () => void };
    if (typeof repo.limpar === "function") {
      repo.limpar();
      await this.repository.persistir();
    }
  }

  private obterOuFalhar(id: string): Task {
    const tarefa = this.repository.buscarPorId(id);
    if (!tarefa) {
      throw new TarefaNaoEncontradaError(id);
    }
    return tarefa;
  }
}

File: legacy/etapa1-python/src/application/task_service.py
"""Camada de aplicação — implementa RF01, RF02 e RF03."""

from __future__ import annotations

from typing import Protocol

from src.domain import Tarefa, TarefaNaoEncontradaError, TituloInvalidoError
from src.domain.task import StatusTarefa


class TaskRepository(Protocol):
    """Contrato de persistência — permite trocar implementação (RNF03)."""

    def salvar(self, tarefa: Tarefa) -> Tarefa: ...

    def listar_todas(self) -> list[Tarefa]: ...

    def buscar_por_id(self, tarefa_id: str) -> Tarefa | None: ...

    def atualizar(self, tarefa: Tarefa) -> Tarefa: ...


class TaskService:
    """
    Serviço de casos de uso da aplicação To-Do.

    Orquestra operações definidas na especificação (UC01–UC03).
    """

    def __init__(self, repository: TaskRepository) -> None:
        self._repository = repository

    def criar_tarefa(self, titulo: str) -> Tarefa:
        """
        RF01 — Cadastrar nova tarefa.

        Raises:
            TituloInvalidoError: quando título é inválido (RN04).
        """
        tarefa = Tarefa.criar(titulo)
        return self._repository.salvar(tarefa)

    def listar_tarefas(self) -> list[Tarefa]:
        """RF02 — Retorna todas as tarefas cadastradas."""
        return self._repository.listar_todas()

    def concluir_tarefa(self, tarefa_id: str) -> Tarefa:
        """
        RF03 — Marca tarefa como concluída.

        Raises:
            TarefaNaoEncontradaError: quando id não existe (RN06).
        """
        tarefa = self._repository.buscar_por_id(tarefa_id)
        if tarefa is None:
            raise TarefaNaoEncontradaError(
                f"Tarefa com id '{tarefa_id}' não encontrada."
            )

        if tarefa.status != StatusTarefa.CONCLUIDA:
            tarefa.concluir()
            return self._repository.atualizar(tarefa)

        return tarefa
